File: java/src/main/java/br/com/emendes/problems/FindFirstAndLastPositionOfElementInSortedArray.java
package br.com.emendes.problems;

import java.util.Arrays;

/**
 * Dada uma matriz de números inteiros classificados em ordem não decrescente,
 * encontre a posição inicial e final de um determinado valor de destino.<br><br>
 *
 * Se o destino não for encontrado na matriz, retorne [-1, -1].<br><br>
 *
 * Você deve escrever um algoritmo com complexidade de tempo de execução O(log n).<br><br>
 *
 * Restrições:<br>
 * 0 <= nums.length <= 10⁵<br>
 * -10⁹ <= nums[i] <= 10⁹<br>
 * nums é uma array não decrescente.<br>
 * -10⁹ <= target <= 10⁹<br>
 */
public class FindFirstAndLastPositionOfElementInSortedArray {

  /**
   * Complexidade de tempo:<br>
   * O(log n) para buscar target no array nums.<br>
   * O(log n) para buscar as ponta esquerda.<br>
   * O(log n) para buscar as ponta direita.<br>
   * Então complexidade de tempo O(3 * log n), logo O(log n).
   */
  public int[] searchRange(int[] nums, int target){
    int numsLength = nums.length;
    int[] answer = {-1, -1};

    if(numsLength == 0) return answer;

    int index = Arrays.binarySearch(nums, target);

    if (index < 0) return answer;

    answer[0] = startsAt(nums, target, index);
    answer[1] = endsAt(nums, target, index);

    return answer;
  }

  /**
   * Busca o índice que contém target mais a esquerda do array nums.
   * @param nums array ordenado.
   * @param target valor a ser buscado.
   * @param seenAt algum índice do array nums que contém target.
   * @return O índice mais a esquerda que contém target.
   */
  private int startsAt(int[] nums, int target, int seenAt) {
    int end = seenAt - 1;
    int start = 0;
    int mid;
    int leftIndex = seenAt;

    while (start <= end) {
      mid = (end - start)/2 + start;
      if (nums[mid] == target) {
        leftIndex = mid;
        end = mid - 1;
      } else {
        start = mid + 1;
      }
    }

    return leftIndex;
  }

  /**
   * Busca o índice que contém target mais a direita do array nums.
   * @param nums array ordenado.
   * @param target valor a ser buscado.
   * @param seenAt algum índice do array nums que contém target.
   * @return O índice mais a direita que contém target.
   */
  private int endsAt(int[] nums, int target, int seenAt) {
    int end = nums.length - 1;
    int start = seenAt + 1;
    int mid;
    int rightIndex = seenAt;

    while (start <= end) {
      mid = (end - start)/2 + start;
      if (nums[mid] == target) {
        rightIndex = mid;
        start = mid + 1;
      } else {
        end = mid - 1;
      }
    }

    return rightIndex;
  }

}

File: javascript/problems/FindFirstAndLastPositionOfElementInSortedArray.js
/**
 * Dado um array de inteiros nums ordenado em ordem crescente, encontre as posições inicial e final de um
 * dado target.
 *
 * Se target não for encontrado no array, retorne [-1, -1].
 *
 * Você deve escrever uma algoritmo que execute com complexidade de tempo O(log n).
 *
 * Restrições:
 *
 * 0 <= nums.length <= 10⁵
 *
 * -10⁹ <= nums[i] <= 10⁹
 *
 * nums está em ordem crescente
 *
 * -10⁹ <= target <= 10⁹
 *
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  let index = binarySearch(nums, target);
  let start = -2;
  let end = 0;
  if (index !== -1) {
    start = index;
    end = index;
    while (start >= 0 && nums[start] === nums[index]) start--;
    while (end < nums.length && nums[end] === nums[index]) end++;
  }

  return [start + 1, end - 1];
};

const binarySearch = (nums, target) => {
  let left = 0;
  let right = nums.length;
  while (left <= right) {
    let mid = Math.floor((left + right) / 2);
    if (nums[mid] === target) return mid;
    if (target > nums[mid]) left = mid + 1;
    else right = mid - 1;
  }

  return -1;
};
